### PR TITLE
feat(transformers): implement initializing deferred libraries

### PR DESCRIPTION
**Pull request by TedSander**
Implement deferred libraries to work with dependency injection and other
angular codegen. This is done by not initializing the library in the parent
ng_deps file when it is declared as deferred, rewriting the import and,
chaining a future that initializes the library in any files that are using
deferred libraries which need angular codegen.

### CODE REVIEW

**Comment by TedSander** [comment]
@kegluneq Can you take a look at this. Apologize in advance for any sloppy code. I probably shouldn't be coding quite this late.

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import 'package:angular2/src/transform/common/asset_reader.dart';
+import 'package:angular2/src/transform/common/logging.dart';
+import 'package:angular2/src/transform/common/names.dart';
+import 'package:barback/barback.dart';
+import 'package:code_transformers/assets.dart';
+import 'package:path/path.dart' as path;
+import 'package:quiver/iterables.dart' as it;
+
+class Rewriter {
+  final String _code;
```
Comment (kegluneq) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: To simplify, could we just read `_code` inside `rewrite` rather than require it as a parameter?

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    visitor.deferredImports.sort(compare);
+    visitor.loadLibraryInvocations.sort(compare);
+
+    var buf = new StringBuffer();
+    var idx = visitor.deferredImports.fold(0,
+        (int lastIdx, ImportDirective node) {
+      buf.write(_code.substring(lastIdx, node.offset));
+
+      var import = _code.substring(node.offset, node.end);
+      buf.write(import.replaceFirst('.dart', '.ng_deps.dart'));
```
Comment (kegluneq) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Reuse `angular2.transform.common.names#toDepsExtension`

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    idx = visitor.loadLibraryInvocations.fold(idx,
+        (int lastIdx, MethodInvocation node) {
+      buf.write(_code.substring(lastIdx, node.offset));
+      var prefix = node.realTarget.name;
+      // Chain a future that initializes the reflector.
+      buf.write('$prefix.loadLibrary().then((_) {$prefix.initReflector();})');
+      return node.end;
+    });
+    if (idx < _code.length) buf.write(_code.substring(idx));
+    return buf.toString();
```
Comment (kegluneq) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: nit: more idiomatic Dart
```
return '$buf';
```

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    super.visitMethodInvocation(node);
+  }
+
+  bool hasDeferredLibrariesToRewrite() {
+    if (deferredImports.isEmpty) {
+      logger.fine('There are no deferred library imports.');
+      return false;
+    }
+    if (loadLibraryInvocations.isEmpty) {
+      logger.fine('There are no loadLibrary invocations that need rewritten.');
```
Comment (kegluneq) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Nit: "need [to be] rewritten"

File: modules/angular2/src/transform/directive_processor/rewriter.dart [hunk 5]
```diff
@@ -106,6 +108,7 @@ class CreateNgDepsVisitor extends Object with SimpleAstVisitor<Object> {
   Object visitImportDirective(ImportDirective node) {
     _maybeWriteImport();
     _updateUsesNonLangLibs(node);
+    if (node.deferredKeyword != null) return null;
```
Comment (kegluneq) [inline modules/angular2/src/transform/directive_processor/rewriter.dart]: Add a comment & code pointer explaining why

File: modules/angular2/test/transform/deferred_rewriter/all_tests.dart [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  _testRewriteDeferredLibraries(
+      'should not rewrite deferred libraries without ng_deps.',
+      'deferred_example_no_ng_deps/index.dart');
+  _testRewriteDeferredLibraries(
+      'should rewrite deferred libraries with ng_deps leave other deferred library alone.',
+      'complex_deferred_example/index.dart');
+}
+
+void _testRewriteDeferredLibraries(String name, String inputPath,
+    {AssetId assetId, AssetReader reader}) {
```
Comment (kegluneq) [inline modules/angular2/test/transform/deferred_rewriter/all_tests.dart]: Looks like these are unused - remove.

File: modules/angular2/test/transform/deferred_rewriter/complex_deferred_example/hello.dart [hunk 7]
```diff
@@ -0,0 +1,10 @@
+library examples.src.hello_world.absolute_url_expression_files;
+
+import 'package:angular2/angular2.dart'
+    show bootstrap, Component, Directive, View, NgElement;
+
+@Component(selector: 'hello-app')
+@View(
+    templateUrl: 'package:other_package/template.html',
+    styleUrls: const ['package:other_package/template.css'])
```
Comment (kegluneq) [inline modules/angular2/test/transform/deferred_rewriter/complex_deferred_example/hello.dart]: Consider removing `styleUrls` -- they seem unrelated and are already tested elsewhere

Comment (TedSander) [inline modules/angular2/test/transform/deferred_rewriter/complex_deferred_example/hello.dart]: Done

File: modules/angular2/test/transform/deferred_rewriter/all_tests.dart (hunk 6 above)
Comment (TedSander) [inline modules/angular2/test/transform/deferred_rewriter/all_tests.dart]: Done

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart (hunk 1 above)
Comment (TedSander) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Done

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart (hunk 2 above)
Comment (TedSander) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Thanks for the tip. I removed _toDepsUri in favor of toDepsExtension. This line happens to be the full import statement. So it wasn't straightforward to use it. Instead I used the constant from the file and kept the replace first.

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart (hunk 3 above)
Comment (TedSander) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Done

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart (hunk 4 above)
Comment (TedSander) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Done

File: modules/angular2/src/transform/directive_processor/rewriter.dart (hunk 5 above)
Comment (TedSander) [inline modules/angular2/src/transform/directive_processor/rewriter.dart]: Done

**Comment by TedSander** [comment]
@kegluneq please take another look. Responded to comments and Fixed some analyzer errors.

File: modules/angular2/src/transform/deferred_rewriter/rewriter.dart (hunk 1 above)
Comment (kegluneq) [inline modules/angular2/src/transform/deferred_rewriter/rewriter.dart]: Good try, but I'm not falling for it

**Comment by kegluneq** [comment]
Other than the `_code` change, lgtm. Thanks, Ted!

**Comment by TedSander** [comment]
Closed by 5cc84ed4bb79e547361b46d61cca4e4efca7b631
